nova_plugin/server.py

```diff
@@ -461,13 +464,250 @@ def stop(nova_client, **kwargs):
     else:
         ctx.logger.info('Server is already stopped')
 
 
+def _server_start(nova_client, server):
+    if server.status != SERVER_STATUS_ACTIVE:
+        nova_client.servers.start(server)
+        # check that we started vm
+        retry = 10
+        while True:
+            server = nova_client.servers.get(server.id)
+            if server.status == SERVER_STATUS_ACTIVE:
+                break
+            ctx.logger.info("{}: Server has {} state."
+                            .format(retry, server.status))
+            time.sleep(30)
+            retry -= 1
+
+            if retry <= 0:
+                ctx.logger.info("Server still has {} state."
+                                .format(server.status))
+                break
+    else:
+        ctx.logger.info('Server is already started?')
+
+
+def _server_suspend(nova_client, server):
+    if server.status == SERVER_STATUS_ACTIVE:
+        nova_client.servers.suspend(server)
+    else:
+        ctx.logger.info('Server is already suspended?')
+
+
+def _server_resume(nova_client, server):
+    if server.status == SERVER_STATUS_SUSPENDED:
+        nova_client.servers.resume(server)
+    else:
+        ctx.logger.info('Server is already resumed?')
+
+
+def _get_snapshot_name(ctx, kwargs):
+    return "vm-{}-{}".format(get_openstack_id(ctx), kwargs["snapshot_name"])
+
+
+def _check_finished_upload(nova_client, server, waiting_list):
+    # check that we created images
+    ctx.logger.info("Check upload state....")
+
+    retry = 10
+    while True:
+        server = nova_client.servers.get(server.id)
+        state = getattr(server, OS_EXT_STS_TASK_STATE)
+        if state not in waiting_list:
+            break
+        ctx.logger.info("{}: Task state: {}."
+                        .format(retry, state))
+        time.sleep(30)
+        retry -= 1
+
+        if retry <= 0:
+            ctx.logger.info("Task still has {} state."
+                            .format(state))
+            break
+
+
+@operation
+@with_nova_client
+def freeze_suspend(nova_client, **kwargs):
+    """
+    Create server backup.
+    """
+    server = get_server_by_context(nova_client)
+    ctx.logger.info("Suspend VM {}".format(server.human_id))
+    _server_suspend(nova_client, server)
+
+
+@operation
+@with_nova_client
+def freeze_resume(nova_client, **kwargs):
+    """
+    Create server backup.
+    """
+    server = get_server_by_context(nova_client)
+    ctx.logger.info("Resume VM {}".format(server.human_id))
+    _server_resume(nova_client, server)
+
+
+@operation
+@with_nova_client
+@with_glance_client
+def snapshot_create(nova_client, glance_client, **kwargs):
+    """
+    Create server backup.
+    """
+    server = get_server_by_context(nova_client)
+
+    ctx.logger.info("Create snapshot for {}".format(server.human_id))
+
+    snapshot_name = _get_snapshot_name(ctx, kwargs)
+    snapshot_rotation = int(kwargs["snapshot_rotation"])
+    snapshot_incremental = kwargs["snapshot_incremental"]
```

Comment:
@0lvin I think we need to double check if these params ``snapshot_rotation`` && ``snapshot_incremental`` does not exists then this will break

Reply:
@mabuaisha We always set this params in case backups, look to https://github.com/cloudify-incubator/cloudify-utilities-plugin/blob/master/cloudify_suspend/workflows.py#L151